src/core/lib/iomgr/parse_address.cc

```diff
@@ -56,6 +56,15 @@ bool grpc_parse_unix(const grpc_uri* uri,
   if (path_len == maxlen) return false;
   un->sun_family = AF_UNIX;
   strcpy(un->sun_path, uri->path);
+  if (un->sun_path[0] == '@') {
```

Comment:
One side-effect of this is that it will basically make it impossible to specify a relative path for a non-abstract socket that begins with an `@` -- anyone who has such a configuration will break, and they would need to work around it by saying something like `./@...`.

I think such a case is unlikely in practice, but I'm not crazy about imposing this kind of limitation in principle.  This seems like a fairly ugly hack.

I've looked around for how other tools and languages deal with this.  It looks like the `@`-prefix approach [is used in golang](https://github.com/golang/go/blob/b2a8317b31d652b3ee293a313269b8290bcdf96c/src/syscall/syscall_linux.go#L393), and I've found [one java library that uses it](https://github.com/kohlschutter/junixsocket/blob/4d3b2af23fb8d1f011dbf7ae01cd51f2d5986b19/junixsocket-demo/src/main/java/org/newsclub/net/unix/demo/DemoHelper.java#L161), but [netty appears to use a leading nul byte](https://github.com/netty/netty/blob/7d971a78a0869171948bb479b3c114d5e772ad45/transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollAbstractDomainSocketEchoTest.java#L26), and [it looks like python does too](https://stackoverflow.com/questions/30818410/can-not-connect-to-an-abstract-unix-socket-in-python).  So it's not clear to me that the '@'-prefix is a universal convention.

Given the lack of a strong convention here, I am not inclined to do it this way.  A cleaner alternative might be to indicate use of an abstract socket via a query param in the `unix:` URI.

I'd like to get feedback from @menghanl and @ejona86 on this, in terms of how this would work in grpc-go and grpc-java.

Reply:
A query param feels like overkill here.  Go's built-in dialer uses the "@" prefix.  It seems extraordinarily unlikely that someone would create a unix socket filename starting with an "@" symbol, especially because of this convention used in many places.  I don't see any problem with following this convention.  If we actually broke anyone I'd be beyond surprised.

Reply:
I'm open to using another convention. Query params would have the same problem, since `?` is a valid character on the unix filesystem.

The most robust thing I can think of would be to indicate it as a separate scheme, i.e. `unix-abstract:path`

Reply:
> Query params would have the same problem, since ? is a valid character on the unix filesystem.

If we are supporting query parameters, a `?` in the path would need to be percent-encoded.

> The most robust thing I can think of would be to indicate it as a separate scheme, i.e. unix-abstract:path

FWIW, we can't change Go's behavior at this point, as we'd be breaking anyone relying upon `unix:@abstract` to work.  We could support an _explicit_ `unix-abstract` scheme if we needed to (we'd just prefix with an `@` ourselves), but I would rather not do this.

Reply:
As Doug said, a `?` character in the path would need to be percent-encoded.  That's already true today, so adding a query param doesn't break anything.  I don't think adding a new URI scheme would actually buy us anything over a query param.

Doug, I think that Go could continue to accept both the query param and the `@`-prefix, for backward compatibility.  And given that the golang libraries themselves use the `@` prefix, it seems reasonable to accept that in Go.  I'm just not sure it's that reasonable in other languages.

Let's see what Eric thinks when he gets back.  We can discuss in more detail next week.

Reply:
@ seems reasonably fair. But I'd feel more comfortable with `@` if I knew its origin and where the convention came from. Can we find any other commonly used system other than Go that uses the convention? It appears it is because ctrl-@ (^@) is NUL. I think using `./@filename` is a fine workaround in the rare case of a name collision, because it should be extremely rare.

Another option would be to %-encode the NUL. That is permitted in URIs. C's URI parser should maybe be disallowing that today (https://tools.ietf.org/html/rfc3986#section-7.3), but Go and Java are probably fine with such things.

From a security perspective, using a different scheme would be best, as these are different namespaces and could cause confusion (both @ and %00). But since anyone concerned with security should be passing an absolute path starting with `/`, the argument seems a bit hollow.

Reply:
I'd guess that the @-prefix convention originally came from the fact that that's how `lsof` prints the names, probably due to ^@ being NUL.

I don't think there is actually a namespace issue with %00, because that's the only character not allowed in a unix path name.  But there is a namespace issue with @.

I would be fine with %-encoding the NUL.  If C-core's parser doesn't currently handle that, we could probably fix it to do so.  If that works for both Java and Go also, let's go with that approach.

If Go already supports @, there's no immediate need to do so; we should just also make sure that %00 works so that people can use the same form in all languages.

@ejona86 and @dfawley, does that work for you?

Reply:
We don't do %-decoding in our target parsing; IIRC we had issues trying to use go's URI.Parse, but I don't remember what they were.  Doing this could break backward compatibility for our users.  We could potentially hack something in where we only handle decoding a percent-encoded NUL as the start of a UDS path(??).  I'd probably rather have a different scheme than that, though.

Reply:
Doug, I think you absolutely need to be doing %-decoding in your target parsing.  As we previously discussed, the target string is an RFC-3986-compliant URI, and %-decoding is required by that RFC.  So I think this is something you need to fix anyway.

Reply:
Making a change to suddenly start %-decoding would affect all our users in a **much** riskier way than adding support for the "@" convention as the first character of a unix path.  We can't make this change; we'd need a different API instead (`NewClient` instead of `Dial`?  I'd love to do this but we don't have the time for it right now).

Reply:
Your current behavior is broken.  I think you should prioritize this change.  The longer you wait before fixing this, the harder it will be.

In any case, for this particular issue, you could special-case `%00` at the start of a `unix:` URI without solving the more general problem.

Reply:
> Your current behavior is broken. I think you should prioritize this change. The longer you wait before fixing this, the harder it will be.

We've *loooong* since reached critical mass on this.  We can't make such a widespread behavior change without doing it in a new API.  On the plus side we also can implement some other gRPC-standard-but-Go-missed-the-memo behavior like: "DNS as default scheme" and "channel begins in IDLE state".  https://github.com/grpc/grpc-go/issues/1786

Back to the issue at hand, it looks like we have 3 options:

1. Support leading-`@` in C and Java
2. New scheme in all languages: `unix-abstract` (e.g.)
3. `%00` as a special case for `unix` scheme in Go, and implement abstract sockets in C/Java if they don't "Just Work" with this encoding

Reply:
Prior art. Null prefix:
* [Linux](https://www.man7.org/linux/man-pages/man7/unix.7.html)
* [Netty](https://github.com/netty/netty/blob/7d971a78a0869171948bb479b3c114d5e772ad45/transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollAbstractDomainSocketEchoTest.java#L26)
* [Python](https://docs.python.org/3/library/socket.html)
* [junixsocket](https://javadoc.io/doc/com.kohlschutter.junixsocket/junixsocket-common/latest/org.newsclub.net.unix/org/newsclub/net/unix/AFUNIXSocketAddress.html) (there is also a convenience factory method that prepends the \0) (mainly including it because it was miscategorized before)

@-prefix:
* [/proc/net/unix](https://man7.org/linux/man-pages/man5/proc.5.html)
* netstat (for display only; probably due to /proc/net/unix)
* [Go](https://github.com/golang/go/blob/333e90448a0e55f2e1161853caecf3d30ef3a74a/src/syscall/syscall_linux.go#L426-L427) ([more](https://github.com/golang/go/blob/333e90448a0e55f2e1161853caecf3d30ef3a74a/src/syscall/syscall_linux.go#L504-L517)). May be incorrectly categorized. Does not appear to be publicly documented, in fact, [goes against "the address must be a file system path"](https://golang.org/pkg/net/#Dial).

Separate scheme/api:
* [socat](https://manpages.debian.org/testing/socat/socat.1.en.html) (UNIX-LISTEN:/some/path vs ABSTRACT-LISTEN:somename)
* [dbus](https://dbus.freedesktop.org/doc/dbus-specification.html#addresses) (unix:path=/some/path vs unix:abstract=somename)
* curl ([--abstract-unix-socket vs --unix-socket](https://curl.haxx.se/docs/manpage.html)) ([CURLOPT_ABSTRACT_UNIX_SOCKET](https://curl.haxx.se/libcurl/c/CURLOPT_ABSTRACT_UNIX_SOCKET.html) vs [CURLOPT_UNIX_SOCKET_PATH](https://curl.haxx.se/libcurl/c/CURLOPT_UNIX_SOCKET_PATH.html))

It appears abstract socket names are fully opaque and can be arbitrary binary, even additional nulls. I don't know if there's any conventions to encode them using the current locale (e.g., UTF-8) or whether they are left uninterpreted (raw bytes). It seems there's a mix in their usage. For example, Python accepts both bytes and strings, but returns the address as bytes. It also seems it doesn't need to be null-terminated.

junixsocket does not use @-prefix. [The referenced code](https://github.com/kohlschutter/junixsocket/blob/4d3b2af23fb8d1f011dbf7ae01cd51f2d5986b19/junixsocket-demo/src/main/java/org/newsclub/net/unix/demo/DemoHelper.java#L161) was just an example. That doesn't make it useless, but greatly diminishes its persuasion power.

I think I'm preferring using a separate scheme. The null-prefix seems like a low-level syscall API encoding and not something that should be user-visible. I'm surprised @-prefix is not more widespread, because it seems users assume "that's the way it is;" I have to assume that is due to netstat's influence. A separate scheme would also be useful as we may want to process the URI as bytes instead of Unicode.

Reply:
> I have to assume that is due to netstat's influence.

I would assume it's "the inability to type a NIL character".

> A separate scheme would also be useful as we may want to process the URI as bytes instead of Unicode.

In this case I would advocate for it to be percent-encoded (without the leading NIL?).  Target strings are not supposed to be treated as raw bytes, and I don't think any scheme should be special in that regard.  In go, we could apply percent-decoding to this new scheme without worrying about backward compatibility.

Reply:
Yeah, whatever we do here has to be RFC-3986-compliant, so it will need to be percent-encoded.

I'd still prefer to trigger this with `%00` with the existing `unix` scheme.  That seems to be the most common API from Eric's list, and it's most consistent with how the underlying syscall API actually works.  (I'd argue that that was a really bad design choice for the syscall API, but that ship has sailed.)

Doug, is there any reason you can't special-case using percent-encoding for the entire path when the path starts with `%00`?

Reply:
> Doug, is there any reason you can't special-case using percent-encoding for the entire path when the path starts with %00?

No, if we decide on `unix`-plus-NIL-prefix and the path starts with `%00`, we could just as easily %-decode the whole string.  For the record I'm fine with any of the 3 options, with no strong feelings about any of them.  They all have similar pros and cons.

Reply:
Okay, then I vote for `%00` prefix with the existing `unix` scheme, and assuming the entire string is percent-encoded.

Eric, any objections?

Reply:
> Okay, then I vote for %00 prefix with the existing unix scheme, and assuming the entire string is percent-encoded.

So it sounds like two votes for a separate scheme, one vote for %00 prefix. I can live with any of the options, but %00 doesn't get my vote.

> Target strings are not supposed to be treated as raw bytes

Any bytes values that are not allowed in that part of the URI would need to be percent encoded. That was not actually the point.

Percent-encoded bytes are generally UTF-8, and some URI implementations may convert them automatically to a Unicode character. But the actual encoding is not defined by URI itself. The point I was making is we may need to process those percent-encoded bytes as bytes and make sure they round-trip as bytes. For example, in Java they will become UTF-16 and bad things will happen. Even in a UTF-8 based language you can't iterate over the runes or check for invalid UTF-8 sequences.

I think C and Go could support raw bytes fairly easily. Java would be harder because strings are UTF-16. It would be easy to update Netty to support bytes. I'm fine with only supporting strings. Although I did notice one abstract socket on my system that appears to be padded with nulls: `@var/run/nvidia-xdriver-673466ff@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@`. Most likely that is a bug.

We should be careful to _not_ null-terminate these strings though; they are length-defined.

> Okay, then I vote for %00 prefix with the existing unix scheme, and assuming the entire string is percent-encoded.

No, we can't assume the entire string is percent-encoded. ASCII letters and numbers will be unencoded. It will actually be quite normal for the entire rest of the string to not need any percent encoding. It seems to be a case where "it's binary but commonly treated like a string."

Reply:
> So it sounds like two votes for a separate scheme, one vote for %00 prefix. I can live with any of the options, but %00 doesn't get my vote.

Who was the second vote for a separate scheme?  I didn't see Doug express a particular preference, although I may have missed that.

If you both prefer a separate scheme, I can live with that.  Let's discuss briefly at today's meeting.

> Although I did notice one abstract socket on my system that appears to be padded with nulls: `@var/run/nvidia-xdriver-673466ff@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@`. Most likely that is a bug.

Yes, I bet that is a bug.  When you look on stackoverflow for questions about abstract sockets, the most common bug is setting the length of the sockaddr wrong, which can result in exactly this kind of thing (or worse, garbage, if you haven't zeroed out the buffer before setting the length wrong).

> > Okay, then I vote for `%00` prefix with the existing unix scheme, and assuming the entire string is percent-encoded.
> 
> No, we can't assume the entire string is percent-encoded. ASCII letters and numbers will be unencoded. It will actually be quite normal for the entire rest of the string to not need any percent encoding. It seems to be a case where "it's binary but commonly treated like a string."

Right, sorry, I worded that statement poorly: I didn't mean that every character in the string had to be percent-encoded; I just meant that we need to percent-decode the entire string, because any character in the string (not just the first one) could be percent-encoded.

Reply:
> Who was the second vote for a separate scheme? I didn't see Doug express a particular preference, although I may have missed that.

I assume he's referencing this:

>> We could potentially hack something in where we only handle decoding a percent-encoded NUL as the start of a UDS path(??). I'd probably rather have a different scheme than that, though.

I do think a separate scheme would be slightly preferable but my level of discomfort with any of the three options is low.  If we do the NUL prefix, I would just think of it as the "`unix:%00` scheme".

Reply:
Eric, Doug, and I discussed this yesterday and agreed that we'd use a new URI scheme called `unix-abstract` for this.  The path name for this URI scheme will be the name of the abstract socket, NOT including the leading NUL byte.  And as per normal URI rules, any character not allowed by RFC-3986 must be percent-encoded.

@beshaya, please make the necessary changes to this PR.  Thanks!